### ISSUE TITLE
Dynamic Portfolio Weights

### ISSUE THREAD

**Issue by firmai** [skfolio/skfolio#68]
```
from skfolio import Portfolio

X = [
    [0.003, -0.001],
    [-0.001, 0.002],
    [0.0015, 0.004],
]

weights = [0.6, 0.4]

portfolio = Portfolio(X=X, weights=weights)

print(portfolio.returns)
>>> array([0.0014, 0.0002, 0.0025])
```

Starting from here, should there not be a weight that allows for 2D input? So instead of a static portfolio, a dynamic one. 

For example, If I rebalance every month, 

X = [
    [0.003, -0.001],
    ...new months
    [-0.001, 0.002],
    [0.0015, 0.004],
....
]

weights = [
    [0.90, 0.10],
    ..new month
    [0.45, 0.55],
    [0.45, 0.55],
....
]

portfolio = Portfolio(X=X, weights=weights)

**Comment by firmai** [comment]
I am just going to add some context, I am making use of the package in a walk forward basis with a certain lookback period.

```
import numpy as np
import pandas as pd
from skfolio import Population, RiskMeasure
from skfolio.optimization import HierarchicalRiskParity
from skfolio.prior import FactorModel
from skfolio.preprocessing import prices_to_returns

def smart_normalize(weights, adjust_fraction=0.05):
    total_weight = weights.sum()
    if np.isclose(total_weight, 1.0, rtol=1e-5):
        return weights

    adjustment = 1.0 - total_weight
    non_zero_weights = weights[weights > 0]
    num_to_adjust = max(2, int(len(non_zero_weights) * adjust_fraction))

    if adjustment > 0:
        # Adjust smallest weights upwards
        stocks_to_adjust = non_zero_weights.nsmallest(num_to_adjust).index
        adjustment_per_stock = adjustment / num_to_adjust
        weights[stocks_to_adjust] += adjustment_per_stock
    else:
        # Adjust largest weights downwards
        stocks_to_adjust = non_zero_weights.nlargest(num_to_adjust).index
        adjustment_per_stock = adjustment / num_to_adjust
        weights[stocks_to_adjust] += adjustment_per_stock

    return weights

df_returns_sample = df_selected
rebalance_period = 30
transaction_cost = 0.1
n_periods = len(df_returns_sample)
assets = df_returns_sample.columns
weights_df = pd.DataFrame(index=df_returns_sample.index, columns=assets)
lookback = 52
threshold = 0.02
persistence = 60

# Initialize with equal weights
current_weights = pd.Series(1/len(assets), index=assets)
last_stable_weights = current_weights.copy()
change_counter = pd.DataFrame(0, index=assets, columns=range(persistence // rebalance_period))

post_adjustment = True

for i in range(lookback, n_periods, rebalance_period):
    # Get the estimation window
    X = df_returns_sample.iloc[i-lookback:i]

    # Calculate the proportion of NaN values for each asset
    nan_proportion = X.isnull().mean()

    # Keep assets with less than 50% NaN values
    valid_assets = nan_proportion[nan_proportion < 0.01].index

    # For these assets, fill NaN with 0
    X_valid = X[valid_assets].fillna(0)

    # Drop assets with all zero values
    valid_assets = X_valid.columns[X_valid.abs().sum() != 0]
    X_valid = X_valid[valid_assets]

    # Adjust current_weights to only include valid assets
    current_weights_valid = current_weights[valid_assets]
    current_weights_valid = smart_normalize(current_weights_valid)

    # Create and fit the HRP model
    model = HierarchicalRiskParity(
        risk_measure=RiskMeasure.CVAR,
        transaction_costs=transaction_cost,
        previous_weights=current_weights_valid
    )

    try:
      model.fit(X_valid)

      # Get new weights for valid assets
      new_weights_valid = pd.Series(model.weights_, index=valid_assets)

      if post_adjustment==True:
          # Apply threshold and persistence logic
          diff = (new_weights_valid - last_stable_weights[valid_assets]).abs()
          mask_exceed = diff >= threshold

          # Shift the change counter and add new column
          change_counter = change_counter.shift(periods=1, axis=1)
          change_counter.iloc[:, 0] = 0  # Initialize the new column with zeros
          change_counter.loc[valid_assets, 0] = mask_exceed  # Update only for valid assets

          # Check if the threshold has been exceeded for the entire persistence period
          persistent_change = (change_counter.sum(axis=1) == change_counter.shape[1]) & change_counter.index.isin(valid_assets)

          # Update weights only for assets with persistent change
          new_weights_valid = np.where(persistent_change[valid_assets], new_weights_valid, last_stable_weights[valid_assets])

          # Update last stable weights for assets with persistent change
          last_stable_weights[valid_assets] = np.where(persistent_change[valid_assets], new_weights_valid, last_stable_weights[valid_assets])

          # Reset change counter for assets that have been updated
          change_counter.loc[persistent_change] = 0

          # Smart normalize the weights
          new_weights_valid = smart_normalize(pd.Series(new_weights_valid, index=valid_assets))

      # Create a full set of weights, setting 0 for assets not in the optimization
      new_weights = pd.Series(0, index=assets)
      new_weights[valid_assets] = new_weights_valid

      # Assign weights for the next rebalance_period days
      for j in range(i, min(i+rebalance_period, n_periods)):
          weights_df.iloc[j] = new_weights

      # Update current weights
      current_weights = new_weights

      # Print some debug information
      print(f"Period {i}:")
      print(f"  Number of valid assets: {len(valid_assets)}")
      print(f"  Assets with non-zero weights: {(new_weights > 0).sum()}")
      print(f"  Max weight: {new_weights.max():.4f}")
      print(f"  Min non-zero weight: {new_weights[new_weights > 0].min():.4f}")
      print(f"  Assets exceeding threshold: {mask_exceed.sum()}")
      print(f"  Assets meeting persistence criteria: {persistent_change.sum()}")
      print("---")

    except:
      print("Data Empty")
      continue

```

**Comment by HugoDelatte** [comment]
Hello, for rebalancing we use `MultiPeriodPortfolio`. You can find the [doc here](https://skfolio.org/user_guide/portfolio.html#multi-period-portfolio) and examples that uses walk forward [here](https://skfolio.org/auto_examples/6_ensemble/plot_1_stacking.html#parameter-tuning).

The paradigm behind `Portfolio` and `MultiPeriodPortfolio` implementation is to have an evaluation that is homogenous to the convex optimization problem we are solving (https://skfolio.org/user_guide/portfolio.html#id2).

Let me know if that helps.

**Comment by firmai** [comment]
I think I understand

pred_hrp = cross_val_predict(
    model_hrp,
    df_select,
    cv=cv,
    n_jobs=-1,
    portfolio_params=dict(name="HRP"),
)

Here we are dynamically updating the model. Thanks I will let you know if I have any questions as I walk through your example.

**Comment by HugoDelatte** [comment]
Correct, in this example, `cv` is a [WalkForward](https://skfolio.org/generated/skfolio.model_selection.WalkForward.html#skfolio.model_selection.WalkForward) cross-validator. The function `cross_val_predict` will `fit` the HRP model on each training set and predict on each related test set. The prediction result of a single test set is a `Portfolio` object and the result of the `cross_val_predict` is a `MultiPeriodPortfolio` containing each predicted `Portfolio`. 

Note that you can change `cv` to more complex cross-validators such as [CombinatorialPurgedCV](https://skfolio.org/generated/skfolio.model_selection.CombinatorialPurgedCV.html) like in [this examples](https://skfolio.org/auto_examples/5_clustering/plot_3_hrp_vs_herc.html#combinatorial-purged-cross-validation).

**Comment by firmai** [comment]
Here is what I think after some tinkering, in the same vein that we can create single period portfolios:

```
from skfolio import Portfolio

X = [
    [0.003, -0.001],
    [-0.001, 0.002],
    [0.0015, 0.004],
]

weights = [0.6, 0.4]

portfolio = Portfolio(X=X, weights=weights)

print(portfolio.returns)
>>> array([0.0014, 0.0002, 0.0025])
```
There should be a way to hand craft multiperiod portfolios `MultiPeriodPortfolio` so that we are able to make use of all the nice built in fuctionalities that you have made available. 

So allow a method to take inputs like this:

X = [
[0.003, -0.001],
...new months
[-0.001, 0.002],
[0.0015, 0.004],
....
]

weights = [
[0.90, 0.10],
..new month
[0.45, 0.55],
[0.45, 0.55],
....
]

**Comment by firmai** [comment]
Actually I didn't think of that,`MultiPeriodPortfolio([Portfolio(X=X[i], weights=weights[i]) for i in len(weights)])` could be a quick fix for future users as opposed to feeding a 2d vector or weights.

**Comment by HugoDelatte** [comment]
Okay, let's keep it this way then
